### PR TITLE
Return directory listings as HTML pages w/ clickable file lists

### LINKED ISSUE #35: Format GET on directory response as HTML
```
Given a directory at $path
When the server receives a GET $path request
Then it should respond with 200
And a Content-Length header set to the length of the body
And a Content-Type of text/html
And a body containing HTML with links to the files in the directory
```

Should pass cob_spec's [DirectoryLinks](https://github.com/8thlight/cob_spec/tree/master/FitNesseRoot/HttpTestSuite/FileServerTestSuite/DirectoryLinks) test

### PR DESCRIPTION
**Pull request by bgerstle**
> Resolves #35

### CODE REVIEW

**Comment by bgerstle** [comment]
@trptcolin Curious what your thoughts are around how cob_spec tests for this. IMHO the way I'm doing it (actually navigating & clicking on files) is much closer to the intended functionality, as opposed to searching for specific elements in the page.

File: src/main/java/com/eighthlight/fabrial/http/request/RequestReader.java [hunk 1]
```diff
@@ -40,27 +39,21 @@ private static RequestBuilder withRequestLine(String requestLine) throws Request
...
+      var firstLine = lineReader.readLine();
+      if (firstLine == null || firstLine.isEmpty()) {
+        return Optional.empty();
+      }
+      return Optional.of(
+          RequestReader.builderWithRequestLine(firstLine)
+                       .withHeaders(new HttpHeaderReader(is).readHeaders())
+                       .withBody(is)
+                       .build()
+      );
```
Comment (trptcolin) [inline src/main/java/com/eighthlight/fabrial/http/request/RequestReader.java]: Cool; this got a lot simpler-looking.

File: src/main/java/com/eighthlight/fabrial/utils/Result.java [hunk 2]
```diff
@@ -88,6 +88,10 @@ public V orElseThrow() throws E {
     return Optional.ofNullable(value).orElseThrow(() -> error);
   }
 
+  public <R extends E> V orElseThrow(Function<E, R> exceptionMapper) throws R {
+    return Optional.ofNullable(value).orElseThrow(() -> exceptionMapper.apply(error));
+  }
+
```
Comment (trptcolin) [inline src/main/java/com/eighthlight/fabrial/utils/Result.java]: I see the compiler is sending out some warning about unchecked casts in this file - if you're sure these are safe (or at least unbroken in this codebase), you could suppress the warnings with `@SuppressWarnings("unchecked")`.

#nonblocker, not added by this PR.

Comment (bgerstle) [reply src/main/java/com/eighthlight/fabrial/utils/Result.java]: I don't think there's a way around the casts here—at least based on my reading of similar libraries (see https://github.com/aol/cyclops)

File: src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java [hunk 3]
```diff
@@ -182,24 +183,48 @@ private Response buildOptionsResponse(Request request, ResponseBuilder builder)
...
                       .stream()
                       .sorted()
-                      .reduce((p1, p2) -> p1 + "," + p2)
-                      .orElse("");
-
-    if (contents.isEmpty()) {
+                      .map(p -> {
+                        // would be nice to append "/" to directories
+                        // dir items must not contain leading slash
+                        // or the <base> element won't work
```
[the remark this answers is not in the document]
Comment (bgerstle) [reply src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java]: > Thinking about this for a bit, I wonder if this limitation could imply the <base> element isn't quite the right fit for this use case.

I don't know enough about that element to say whether or not it's the "right" way to use it, but it seems to have the desired effect in my case. In any event, it shouldn't be too hard to just use the full path for reach file.

[the remark this answers is not in the document]
Comment (bgerstle) [reply src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java]: > Accepting trailing / in requests makes sense to me

Sounds like there's a misunderstanding here. My comment from before about cob_spec is that it doesn't allow **leading** slashes in the link's `href` attribute. Having trailing slashes for directories is just a throwaway comment on how I'd prefer the "UI" to look (to distinguish files & directories).

File: src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java [hunk 4]
```diff
@@ -182,24 +183,48 @@ private Response buildOptionsResponse(Request request, ResponseBuilder builder)
...
+    var basePath = basePathBuilder.toString();
+    var contents =
+        String.join(
+            "\n",
+            "<head>",
+            "<base target=_self href=\"" + basePath + "\"/>",
+            "</head>",
+            "<body>",
+            body,
+            "</body>");
```
[the remark this answers is not in the document]
Comment (bgerstle) [reply src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java]: > This HTML generation seems like a natural place for a seam...

Definitely something I'd like to refactor once I've confirmed that it's up & running.

> One potential design for this specific thing: a class that takes whatever base-directory config it needs in a constructor, with a method that takes a list of files and returns an HTML page as a string

The "feature requirements" here are pretty basic: generate an HTML page with a link for each file in the given directory. This could be (and maybe was) a couple-liner stream transform. Most of the other stuff is encapsulating HTML (head/body/etc.) so I could have a spot to put the `<base>` element. 

> I'll make another unit testing plug here too...

Fair, I probably should have considered using an "offline" HTML parser to test the HTML generation. However, I was less worried about getting that right than the actual navigation of the HTML in a browser (hence my recent addition of Selenium-driven tests). 

I'll see if I can pick any low-hanging fruit to speed up test execution.

File: src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (trptcolin) [reply src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java]: Oh no, I thought I did understand! Let me try again, and let me know if I'm still off (and maybe we can jump on a call if so).

I did understand [I think] that there are 2 separate question: prepending a [leading] slash and appending a [trailing] slash, and I was trying to address the trailing slash in the last paragraph of my comment above. Maybe that hurt more than it helped?

On the leading slash, I saw this in the PR description:

> The "Directory Links" cob_spec test fails—but only because it searches for href == /filename and my server uses href == filename.

and I assumed that meant "something different needs to happen to pass cob_spec" which led me to suggest using absolute paths. But in your last comment here you're saying

> My comment from before about cob_spec is that it doesn't allow leading slashes in the link's href attribute

which doesn't sound right to me, because it seems to contradict the PR description quote above, and also the [cob_spec selector that checks the links](https://github.com/8thlight/cob_spec/blob/db6a99ad3f29768f2184ac6c350adfa0e722d07c/src/main/java/HttpBrowser.java#L154). 

I'm thinking that maybe the `href == /filename` pseudocode might be abstract enough that it's causing my misunderstanding of what you mean, but in any case hopefully this comment makes more sense! 😄

[the remark this answers is not in the document]
Comment (bgerstle) [reply src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java]: > But in your last comment here you're saying

Ack, that was a mistake. Other way around: cob_spec expects leading slashes, but I can't use them since it breaks the "relative path" behavior I get with `<base>`.

So you were right. I didn't fully grok your comment about updating cob_spec to accept trailing slashes, which I then mistakenly thought was somehow referencing my main issue with leading slashes causing cob_spec failures...

Ultimately, this is just another reason why I shouldn't leave "feature requests" for cob_spec in the comments of my server 😓 ...

**Comment by trptcolin** [comment]
One other note, though you're probably already looking into it - CI looks like it's failing on an `Optional` unwrapping error in `com.eighthlight.fabrial.test.http.HttpRequestLineParsingTests > emptyStringTest()`. At least for me, the log output makes it hard to find what exactly is failing - I needed to scroll around a bit to find it. I'd still advocate for sending log output to build artifacts, so they're still findable to debug issues, but keep the build output neat and fast to interpret.

**Comment by bgerstle** [comment]
> I'd still advocate for sending log output to build artifacts, so they're still findable to debug issues, but keep the build output neat and fast to interpret.

I checked, Travis doesn't have a way to store artifacts "out of the box." I'd have to hook it up to S3 or similar. Agreed that the log input is noisy, and I'm trying to catch the noisier exceptions (e.g. empty request "errors") to curtail it. Yay for handling specific exceptions in Java...

File: src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java [hunk 5]
```diff
@@ -187,13 +188,13 @@ private Response buildGetDirectoryResponse(Request request, ResponseBuilder buil
         fileController.getDirectoryContents(request.uri.getPath())
                       .stream()
                       .sorted()
-                      .map(p -> {
```
Comment (bgerstle) [inline src/main/java/com/eighthlight/fabrial/http/file/FileHttpResponder.java]: @trptcolin The "HTML generation" has now been reduced to just a few lines. I've also `@Disabled` most of the (mostly obsolete) `AppAcceptanceTest` cases which should shave a good ~20s off the test execution time.

**Comment by bgerstle** [comment]
🎉 
image

File: src/main/java/com/eighthlight/fabrial/http/request/RequestReader.java (hunk 1 above)
Comment (bgerstle) [reply src/main/java/com/eighthlight/fabrial/http/request/RequestReader.java]: Yeah, as much as I like my `Result` class, I need to be wary of using it "just because." 😝

**Comment by bgerstle** [comment]
@trptcolin just to recap, there are two tests which cover the "directory listing" functionality:

- `FileHttpResponderIntegrationTest.returnsAllFilesInDirWithMultipleFiles` which is a simplistic unit test which makes sure that the response contains the filenames of the directory's children
- `GetDirContentsAcceptanceTest.*` acceptance tests which verify that you can view directory contents, with clickable elements that support nested files & directories.
